File: rider-location-service/src/main/java/com/potatorider/service/RiderLocationService.java
package com.potatorider.service;

import com.potatorider.domain.RiderLocation;
import com.potatorider.repository.DeliveryRepository;
import com.potatorider.repository.RiderLocationRepository;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;

import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class RiderLocationService {

    private final RiderLocationRepository riderLocationRepository;
    private final DeliveryRepository deliveryRepository;

    public Mono<Boolean> tryPutOperation(final RiderLocation riderLocation) {
        return riderLocationRepository
                .setIfPresent(riderLocation)
                .flatMap(isSaved -> orElseSetNew(isSaved, riderLocation));
    }

    private Mono<Boolean> orElseSetNew(final Boolean isSaved, final RiderLocation riderLocation) {
        return isSaved ? Mono.just(true) : doSetNew(riderLocation);
    }

    private Mono<Boolean> doSetNew(final RiderLocation riderLocation) {
        return deliveryRepository
                .isPickedUp(riderLocation.getDeliveryId())
                .flatMap(isPickedUp -> setIfPickedUp(isPickedUp, riderLocation));
    }

    private Mono<Boolean> setIfPickedUp(
            final Boolean isPickedUp, final RiderLocation riderLocation) {
        return isPickedUp ? riderLocationRepository.setIfAbsent(riderLocation) : Mono.just(false);
    }

    public Mono<RiderLocation> getLocation(final String locationId) {
        return riderLocationRepository.getLocation(locationId);
    }
}

File: rider-location-service/src/main/java/com/potatorider/service/RiderLocationService.kt
package com.potatorider.service

import com.potatorider.domain.RiderLocation
import com.potatorider.repository.DeliveryRepository
import com.potatorider.repository.RiderLocationRepository
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class RiderLocationService(
    private val riderLocationRepository: RiderLocationRepository,
    private val deliveryRepository: DeliveryRepository
) {

    fun tryPutOperation(riderLocation: RiderLocation): Mono<Boolean> = riderLocationRepository
        .setIfPresent(riderLocation)
        .flatMap { isSaved -> orElseSetNew(isSaved, riderLocation) }

    private fun orElseSetNew(isSaved: Boolean, riderLocation: RiderLocation): Mono<Boolean> =
        if (isSaved) Mono.just(true) else doSetNew(riderLocation)

    private fun doSetNew(riderLocation: RiderLocation): Mono<Boolean> = deliveryRepository
        .isPickedUp(riderLocation.deliveryId!!)
        .flatMap { isPickedUp -> setIfPickedUp(isPickedUp, riderLocation) }

    private fun setIfPickedUp(
        isPickedUp: Boolean, riderLocation: RiderLocation
    ): Mono<Boolean> =
        if (isPickedUp) riderLocationRepository.setIfAbsent(riderLocation) else Mono.just(false)

    fun getLocation(locationId: String): Mono<RiderLocation> =
        riderLocationRepository.getLocation(locationId)
}
